velox/serializers/PrestoSerializer.cpp

```diff
@@ -1706,8 +1752,37 @@ static inline int32_t rangesTotalSize(
   return total;
 }
 
+template <TypeKind Kind>
+void serializeDictionaryVector(
+    const BaseVector* vector,
+    const folly::Range<const IndexRange*>& ranges,
+    VectorStream* stream) {
+  // Cannot serialize dictionary as PrestoPage dictionary if it has nulls.
+  // Also check if the stream was set up for dictionary (we had to know the
+  // encoding type when creating VectorStream for that).
+  if (vector->nulls() || !stream->isDictionaryStream()) {
+    serializeWrapped(vector, ranges, stream);
+    return;
+  }
+
+  using T = typename KindToFlatVector<Kind>::WrapperType;
+  auto dictionaryVector = dynamic_cast<const DictionaryVector<T>*>(vector);
+
+  std::vector<IndexRange> childRanges;
+  childRanges.push_back({0, dictionaryVector->valueVector()->size()});
```

Comment:
dictionaryVector may contain lots unreferenced values; users need to know what they are doing when using this API.

Reply:
Yes, the API will not re-encode the data and will trust the user supplied data is efficient. This is something we can tackle if we decided to change this API to be smarter about choosing the best way to encode the data.